src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets

```diff
@@ -110,6 +111,9 @@ The .NET Foundation licenses this file to you under the MIT license.
       <LinkerArg Include="-Wl,-z,now" Condition="'$(TargetOS)' != 'OSX'" />
       <LinkerArg Include="-Wl,-u,_NativeAOT_StaticInitialization" Condition="'$(TargetOS)' == 'OSX' and '$(NativeLib)' == 'Shared'" />
       <LinkerArg Include="-Wl,--require-defined,NativeAOT_StaticInitialization" Condition="'$(TargetOS)' != 'OSX' and '$(NativeLib)' == 'Shared'" />
+      <LinkerArg Include="-fuse-ld=lld" Condition="'$(UseLLVMLinker)' == 'true'" />
```

Comment:
This sounds like a workaround for a bug in the objwriter where we are missing some sort of annotations.

What do we need to do to make this workaround unnecessary?

Reply:
This is only selecting llvm's linker on-demand because we didn't installed aarch64 binutils in latest image (we can install that). In the coreclr build, we do this automatically: https://github.com/dotnet/runtime/blob/e90b0071089d655445231c8d4d4b74b5e4fc9030/eng/common/native/init-compiler.sh#L138

or

Did you mean the following lines; `-Wl,-z,nostart-stop-gc` workaround? If so, that is because it breaks the build with clang 13 onwards on Linux. The better fix for that is we annotate `__start_modules`, `__stop_modules` and other section with `__attribute__((retain))` as described here: https://lld.llvm.org/ELF/start-stop-gc. The newer versions of both clang and gcc require that workaround and will benefit from that fix.

Reply:
> we didn't installed aarch64 binutils in latest image (we can install that).

I think it would be better to install binutils so that the testing matches the setup that we recommend end users to use.

> The better fix for that is we annotate __start_modules, __stop_modules and other section with __attribute__((retain))

Yes, we should do that.

Reply:
> Yes, we should do that.

Done.

> the testing matches the setup that we recommend end users to use.

I have used `UseLLVMLinker` only in one platform with latest llvm version. Our Alpine Linux [prerequisites](https://github.com/dotnet/docs/blob/77bc038e2b637a39f5bbdeb656236256699ad8ad/docs/core/deploying/native-aot/index.md#prerequisites) explicitly require `lld`.

Reply:
> Yes, we should do that.

Done.

> the testing matches the setup that we recommend end users to use.

I have used llvm linker only in one platform (`linux-musl-arm64`) with latest llvm toolchain. Our Alpine Linux [prerequisites](https://github.com/dotnet/docs/blob/77bc038e2b637a39f5bbdeb656236256699ad8ad/docs/core/deploying/native-aot/index.md#prerequisites) explicitly mention `lld`, so while it is not the default linker, it is not deviating from the recommendations.  In coreclr, we prefer llvm linker (lld) over gnu one (ld) and automatically select lld when clang is used and lld is available. More details about this preference are in https://github.com/dotnet/runtime/pull/52244.